File: src/main/java/edu/isi/gaia/ScalingTest.java
package edu.isi.gaia;

import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import com.google.common.collect.ImmutableSet;
import org.apache.jena.query.Dataset;
import org.apache.jena.rdf.model.Model;
import org.apache.jena.rdf.model.Resource;
import org.apache.jena.riot.RDFDataMgr;
import org.apache.jena.riot.RDFFormat;
import org.apache.jena.tdb.TDBFactory;
import org.apache.jena.vocabulary.RDF;
import org.apache.jena.vocabulary.SKOS;
import org.apache.jena.vocabulary.XSD;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import static edu.isi.gaia.AIFUtils.*;

public class ScalingTest {

    private static final String LDC_NS = "https://tac.nist.gov/tracks/SM-KBP/2018/LdcAnnotations#";
    private Model model;
    private Resource system;

    // Beginning sizes of data, about what is in T101
    //private int entityCount = 128000;
    //private int eventCount = 38400;

    private int entityCount = 1000;
    private int eventCount = 300;

    private int entityIndex = 1;
    private int eventIndex = 1;
    private int relationIndex = 1;
    private int assertionIndex = 1;

    // Utility values, so that we can easily create random things
    final static String abc = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    final Resource[] entityTypes = SeedlingOntologyMapper.ENTITY_TYPES.toArray(new Resource[0]);

    final SeedlingOntologyMapper ontologyMapping = new SeedlingOntologyMapper();

    private final String putinDocumentEntityUri = getUri("E781167.00398");
    private final String putinElectedDocumentEventUri = getUri("V779961.00010");
    private final String russiaDocumentEntityUri = getUri("E779954.00004");

    private final Random r = new Random();
    private List<Resource> entityResourceList = new ArrayList<>();

    private final String filename = "scalingdata.ttl";

    /**
     * Main function.  Call with no arguments
     */
    public static void main(String[] args) {
        ScalingTest scalingTest = new ScalingTest();
        scalingTest.runtest();
    }

    protected void runtest() {

        // prevent too much logging from obscuring the Turtle examples which will be printed
        ((Logger) org.slf4j.LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME)).setLevel(Level.INFO);

        for (int ii = 0; ii < 200; ii++) {
            System.out.println("Trying :  Entity count: " + entityCount);
            long startTime = System.currentTimeMillis();

            runSingleTest();

            long endTime = System.currentTimeMillis();
            long duration = (endTime - startTime) / 1000;

            long size = 0;
            File f = new File(filename);
            if (f.exists()) {
                size = f.length();
            }
            size /= 1000000.;
            System.out.println(" Size of output: " + size + "  Time (sec) " + duration);

            increase();
        }
    }

    private void increase() {
        entityCount *= 2;
        eventCount *= 2;
    }

    private void runSingleTest() {
        setup();

        for (int ii = 0; ii < entityCount; ii++) {
            addEntity();
        }
        for (int ii = 0; ii < eventCount; ii++) {
            addEvent();
        }

        dumpAndAssertValid(filename);
    }

    private void setup() {
        model = createModel();
        system = makeSystemWithURI(model, getTestSystemUri());
    }

    private void addEntity() {

        // Add an entity
        Resource entityResource = makeEntity(model, getEntityUri(), system);
        entityResourceList.add(entityResource);

        // sometimes add hasName, textValue, or numericValue, NOTE:  This does not check type!!!!
        double rand = r.nextDouble();
        if (rand < 0.15) {
            markName(entityResource, getRandomString(5));
        } else if (rand < 0.3) {
            markTextValue(entityResource, getRandomString(7));
        } else if (rand < 0.4) {
            markNumericValueAsDouble(entityResource, r.nextDouble());
        }

        // Set the type
        Resource typeToUse = entityTypes[r.nextInt(entityTypes.length)];
        Resource typeAssertion = markType(model, getAssertionUri(), entityResource,
                typeToUse, system, 1.0);

        addJustificationAndPrivateData(typeAssertion);
    }

    private void addEvent() {
        // Add an event
        Resource eventResource = makeEvent(model, getEventUri(), system);

        // Set the type
        String eventTypeString = EVENT_TYPES[r.nextInt(EVENT_TYPES.length)];
        Resource typeResource = ontologyMapping.eventType(eventTypeString);
        Resource typeAssertion = markType(model, getAssertionUri(), eventResource, typeResource, system, 1.0);

        addJustificationAndPrivateData(typeAssertion);

        // Make two arguments
        for (int ii = 0; ii < 2; ii++) {
            Resource argument = markAsArgument(model, eventResource,
                    ontologyMapping.eventArgumentTypeNotLowercase(eventTypeString + getRandomSuffix()),
                    getRandomEntity(), system, 0.785, getAssertionUri());
            addJustificationAndPrivateData(argument);
        }
    }

    private void addJustificationAndPrivateData(Resource resource) {
        String docId = getRandomDocId();

        // Justify the type assertion
        markTextJustification(model, ImmutableSet.of(resource), docId, 1029, 1033, system, 0.973);

        // Add some private data
        markPrivateData(model, resource, "{ 'provenance' : '" + docId + "' }", system);
    }

    // we dump the test name and the model in Turtle format so that whenever the user
    // runs the tests, they will also get the examples
    private void dumpAndAssertValid(String testName) {
        try {
            // RDFDataMgr.write(System.out, model, RDFFormat.TURTLE_PRETTY);
            RDFDataMgr.write(Files.newOutputStream(Paths.get(testName)), model, RDFFormat.TURTLE_PRETTY);
        } catch (Exception e) {
            System.err.println("Unable to write to file " + testName + " " + e.getMessage());
            e.printStackTrace();
        }
    }

    private Model createModel() {

        // Make a disk model
        Dataset dataset = TDBFactory.createDataset("/tmp/model");
        Model model = dataset.getDefaultModel();

        // final Model model = ModelFactory.createDefaultModel();
        // adding namespace prefixes makes the Turtle output more readable
        model.setNsPrefix("rdf", RDF.uri);
        model.setNsPrefix("xsd", XSD.getURI());
        model.setNsPrefix("aida", AidaAnnotationOntology.NAMESPACE);
        model.setNsPrefix("ldcOnt", SeedlingOntologyMapper.NAMESPACE_STATIC);
        model.setNsPrefix("ldc", LDC_NS);
        model.setNsPrefix("skos", SKOS.uri);
        return model;
    }

    private static String getUri(String localName) {
        return LDC_NS + localName;
    }

    private String getEntityUri() {
        return getUri("entity-" + entityIndex++);
    }

    private String getEventUri() {
        return getUri("event-" + eventIndex++);
    }

    private String getRelationUri() {
        return getUri("relation-" + relationIndex++);
    }

    private String getAssertionUri() {
        return getUri("assertion-" + assertionIndex++);
    }

    private String getTestSystemUri() {
        return getUri("testSystem");
    }

    private String getRandomDocId() {
        String s = "";
        if (r.nextBoolean()) {
            s += "IC";
        } else {
            s += "HC";
        }
        s += "00";
        s += "" + (r.nextInt(1000));
        s += abc.charAt(r.nextInt(abc.length()));
        s += abc.charAt(r.nextInt(abc.length()));
        s += abc.charAt(r.nextInt(abc.length()));
        return s;
    }

    private String getRandomString(int length) {
        String s = "";
        for (int ii = 0; ii < length; ii++) {
            s += abc.charAt(r.nextInt(abc.length()));
        }
        return s;
    }

    private Resource getRandomEntity() {
        return entityResourceList.get(r.nextInt(entityResourceList.size()));
    }


    /**
     * Return a string representing an argument
     */
    private String getRandomSuffix() {
        String s = "_" + ROLES[r.nextInt(ROLES.length)];
        return s;
    }


    private final String[] EVENT_TYPES = {
            "Business.DeclareBankruptcy", "Business.End", "Business.Merge", "Business.Start",
            "Conflict.Attack", "Conflict.Demonstrate",
            "Contact.Broadcast", "Contact.Contact", "Contact.Correspondence", "Contact.Meet",
            "Existence.DamageDestroy",
            "Government.Agreements", "Government.Legislate", "Government.Spy", "Government.Vote",
            "Inspection.Artifact", "Inspection.People",
            "Justice.Acquit", "Justice.Appeal", "Justice.ArrestJail", "Justice.ChargeIndict", "Justice.Convict",
            "Justice.Execute", "Justice.Extradite", "Justice.Fine", "Justice.Investigate", "Justice.Pardon",
            "Justice.ReleaseParole", "Justice.Sentence", "Justice.Sue", "Justice.TrialHearing",
            "Life.BeBorn", "Life.Die", "Life.Divorce", "Life.Injure", "Life.Marry",
            "Manufacture.Artifact",
            "Movement.TransportArtifact", "Movement.TransportPerson",
            "Personnel.Elect", "Personnel.EndPosition", "Personnel.Nominate", "Personnel.StartPosition",
            "Transaction.Transaction", "Transaction.TransferControl", "Transaction.TransferMoney",
            "Transaction.TransferOwnership"};

    private final String[] ROLES = {"Attacker", "Instrument", "Place", "Target", "Time", "Broadcaster",
            "Place", "Time", "Participant", "Place", "Participant", "Time",
            "Participant", "Affiliate", "Affiliation", "Affiliation", "Person",
            "Entity", "Sponsor", "Defendant", "Prosecutor", "Adjudicator",
            "Defendant", "Agent", "Instrument", "Victim", "Artifact",
            "Manufacturer", "Agent", "Artifact", "Destination", "Instrument",
            "Origin", "Time", "Agent", "Destination", "Instrument", "Origin",
            "Person", "Employee", "Organization", "Person", "Entity", "Place",
            "Beneficiary", "Giver", "Recipient", "Thing", "Time"};

}

File: python/tests/scaling_test.py
import sys
import os
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../')))
import time
import random

from aida_interchange import aifutils
from aida_interchange.rdf_ontologies import seedling_ontology
from rdflib import URIRef


class ScalingTest():
    filename = "scalingdata.ttl"
    LDC_NS = "https://github.com/NextCenturyCorporation/AIDA-Interchange-Format/LdcAnnotations#"
    g = aifutils.make_graph()
    system = aifutils.make_system_with_uri(g, 'http://www.test.edu/testSystem')

    # beginning sizes of data
    entity_count = 128000
    event_count = 38400
    relations_count = 200
    assertion_count = 1500

    entity_index = 1
    event_index = 1
    relation_index = 1
    assertion_index = 1

    # utility values, so taht we can easily create random things
    abc = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

    entity_resource_list = []


    def run_scaling_test(self):
        for ii in range(200):
            print("trying : Entity count: ", self.entity_count)
            start_time = int(round(time.time() * 1000))

            self.run_single_test()

            end_time = int(round(time.time() * 1000))
            duration = (start_time - end_time) / 1000
            size = 0

            f = open(self.filename)
            if os.path.isfile(self.filename):
                size = os.path.getsize(self.filename)
            size /= 1000000.
            print("Size of output: ", size, " duration: ", duration)

            # double size of entities and events after every iteration
            self.increase()


    def run_single_test(self):
        # adds entities and events and wrties to file
        for ii in range(self.entity_count):
            self.add_entity()

        for i in range(self.event_count):
            self.add_event()

        self.write_to_file(self.filename)


    def add_entity(self):
        # add an entity
        entity_resource = aifutils.make_entity(self.g, self.get_entity_uri(), self.system)
        self.entity_resource_list.append(entity_resource)

        # sometimes add hasName, textValue, or numericValue, NOTE: This does not check type!!!
        rand = random.random()
        if rand < 0.15:
            aifutils.mark_name(self.g, entity_resource, self.get_random_string(5))
        elif rand < 0.3:
            aifutils.mark_text_value(self.g, entity_resource, self.get_random_string(7))
        elif rand < 0.4:
            aifutils.mark_numeric_value_as_double(self.g, entity_resource, random.random())

        # set the type
        type_to_use =  self.get_random_entity()
        type_assertion = aifutils.mark_type(self.g, self.get_assertion_uri(), entity_resource, type_to_use, self.system, 1.0)
        self.add_justification_and_private_data(type_assertion)


    def add_event(self):
        # add an event
        event_resource = aifutils.make_event(self.g, self.get_event_uri(), self.system)

        # add the type
        type_resource = self.EVENT_TYPES[random.randint(0, len(self.EVENT_TYPES)) - 1]
        type_assertion = aifutils.mark_type(self.g, self.get_assertion_uri(), event_resource, type_resource, self.system, 1.0)

        self.add_justification_and_private_data(type_assertion)

        # make two arguments
        for i in range(2):
            arg = URIRef(type_resource + self.get_random_suffix())
            argument = aifutils.mark_as_argument(self.g, event_resource, arg, self.get_random_entity(), self.system,
                                                                         0.785, self.get_assertion_uri())
            self.add_justification_and_private_data(argument)


    def add_justification_and_private_data(self, resource):
        docId = self.get_random_doc_id()

        # justify the type assertion
        aifutils.mark_text_justification(self.g, resource, docId, 1029, 1033, self.system, 0.973)

        # add some private data
        aifutils.mark_private_data(self.g, resource, "{ 'provenance' : '" + docId + "' }", self.system)


    def increase(self):
        self.entity_count *= 2
        self.event_count *= 2


    def get_uri(self, uri):
        return self.LDC_NS + uri


    def get_entity_uri(self):
        self.entity_index += 1
        return self.get_uri("entity-" + str(self.entity_index))


    def get_event_uri(self):
        self.event_index += 1
        return self.get_uri("event-" + str(self.event_index))


    def get_relation_uri(self):
        self.relation_index += 1
        return self.get_uri("relation-" + str(self.relation_index))


    def get_assertion_uri(self):
        self.assertion_index += 1
        return self.get_uri("assertion-" + str(self.assertion_index))


    def get_test_system_uri(self):
        return self.get_uri("testSystem")


    def get_random_doc_id(self):
        s = ""
        if random.getrandbits(1) == 1:
            s += "IC"
        else:
            s += "HC"
        s += "00"
        s += "" + str((random.randint(0, 1000)))
        s += self.abc[random.randint(0, len(self.abc) - 1)]
        s += self.abc[random.randint(0, len(self.abc) - 1)]
        s += self.abc[random.randint(0, len(self.abc) - 1)]
        return s


    def get_random_string(self, length):
        s = ""
        for i in range(0, length):
            s += self.abc[random.randint(0, len(self.abc) - 1)]
        return s


    def get_random_entity(self):
        return self.ENTITY_TYPES[random.randint(0, len(self.ENTITY_TYPES) - 1)]


    def get_random_suffix(self):
        s = "_" + self.ROLES[random.randint(0, len(self.ROLES) - 1)]
        return s


    def write_to_file(self, testname):
        print("\n\n", testname, "\n\n")
        file = open(testname, "w")
        file.write(str(self.g.serialize(format='turtle')))
        file.close()

    ENTITY_TYPES = [
        seedling_ontology.Person,
        seedling_ontology.Organization,
        seedling_ontology.Location,
        seedling_ontology.Facility,
        seedling_ontology.GeopoliticalEntity,
        seedling_ontology.FillerType,
        seedling_ontology.Business_DeclareBankruptcy,
        seedling_ontology.Business_End,
        seedling_ontology.Business_Merge,
        seedling_ontology.Business_Start,
        seedling_ontology.Conflict_Attack,
        seedling_ontology.Conflict_Demonstrate,
        seedling_ontology.Contact_Broadcast,
        seedling_ontology.Contact_Contact,
        seedling_ontology.Contact_Correspondence,
        seedling_ontology.Contact_Meet,
        seedling_ontology.Existence_DamageDestroy,
        seedling_ontology.Government_Agreements,
        seedling_ontology.Government_Legislate,
        seedling_ontology.Government_Spy,
        seedling_ontology.Government_Vote,
        seedling_ontology.Inspection_Artifact,
        seedling_ontology.Inspection_People,
        seedling_ontology.Justice_Acquit,
        seedling_ontology.Justice_Appeal,
        seedling_ontology.Justice_ArrestJail,
        seedling_ontology.Justice_ChargeIndict,
        seedling_ontology.Justice_Convict,
        seedling_ontology.Justice_Execute,
        seedling_ontology.Justice_Extradite,
        seedling_ontology.Justice_Fine,
        seedling_ontology.Justice_Investigate,
        seedling_ontology.Justice_Pardon,
        seedling_ontology.Justice_ReleaseParole,
        seedling_ontology.Justice_Sentence,
        seedling_ontology.Justice_Sue,
        seedling_ontology.Justice_TrialHearing,
        seedling_ontology.Life_BeBorn,
        seedling_ontology.Life_Die,
        seedling_ontology.Life_Divorce,
        seedling_ontology.Life_Injure,
        seedling_ontology.Life_Marry,
        seedling_ontology.Manufacture_Artifact,
        seedling_ontology.Movement_TransportArtifact,
        seedling_ontology.Movement_TransportPerson,
        seedling_ontology.Personnel_Elect,
        seedling_ontology.Personnel_EndPosition,
        seedling_ontology.Personnel_Nominate,
        seedling_ontology.Personnel_StartPosition,
        seedling_ontology.Transaction_Transaction,
        seedling_ontology.Transaction_TransferControl,
        seedling_ontology.Transaction_TransferMoney,
        seedling_ontology.Transaction_TransferOwnership,
        seedling_ontology.GeneralAffiliation_APORA,
        seedling_ontology.GeneralAffiliation_MORE,
        seedling_ontology.GeneralAffiliation_OPRA,
        seedling_ontology.GeneralAffiliation_OrganizationWebsite,
        seedling_ontology.GeneralAffiliation_PersonAge,
        seedling_ontology.GeneralAffiliation_Sponsorship,
        seedling_ontology.Measurement_Count,
        seedling_ontology.OrganizationAffiliation_EmploymentMembership,
        seedling_ontology.OrganizationAffiliation_Founder,
        seedling_ontology.OrganizationAffiliation_InvestorShareholder,
        seedling_ontology.OrganizationAffiliation_Leadership,
        seedling_ontology.OrganizationAffiliation_Ownership,
        seedling_ontology.OrganizationAffiliation_StudentAlum,
        seedling_ontology.PartWhole_Membership,
        seedling_ontology.PartWhole_Subsidiary,
        seedling_ontology.PersonalSocial_Business,
        seedling_ontology.PersonalSocial_Family,
        seedling_ontology.PersonalSocial_RoleTitle,
        seedling_ontology.PersonalSocial_Unspecified,
        seedling_ontology.Physical_LocatedNear,
        seedling_ontology.Physical_OrganizationHeadquarter,
        seedling_ontology.Physical_OrganizationLocationOrigin,
        seedling_ontology.Physical_Resident]

    EVENT_TYPES = [
        seedling_ontology.Business_DeclareBankruptcy,
        seedling_ontology.Business_End,
        seedling_ontology.Business_Merge,
        seedling_ontology.Business_Start,
        seedling_ontology.Conflict_Attack,
        seedling_ontology.Conflict_Demonstrate,
        seedling_ontology.Contact_Broadcast,
        seedling_ontology.Contact_Contact,
        seedling_ontology.Contact_Correspondence,
        seedling_ontology.Contact_Meet,
        seedling_ontology.Existence_DamageDestroy,
        seedling_ontology.Government_Agreements,
        seedling_ontology.Government_Legislate,
        seedling_ontology.Government_Spy,
        seedling_ontology.Government_Vote,
        seedling_ontology.Inspection_Artifact,
        seedling_ontology.Inspection_People,
        seedling_ontology.Justice_Acquit,
        seedling_ontology.Justice_Appeal,
        seedling_ontology.Justice_ArrestJail,
        seedling_ontology.Justice_ChargeIndict,
        seedling_ontology.Justice_Convict,
        seedling_ontology.Justice_Execute,
        seedling_ontology.Justice_Extradite,
        seedling_ontology.Justice_Fine,
        seedling_ontology.Justice_Investigate,
        seedling_ontology.Justice_Pardon,
        seedling_ontology.Justice_ReleaseParole,
        seedling_ontology.Justice_Sentence,
        seedling_ontology.Justice_Sue,
        seedling_ontology.Justice_TrialHearing,
        seedling_ontology.Life_BeBorn,
        seedling_ontology.Life_Die,
        seedling_ontology.Life_Divorce,
        seedling_ontology.Life_Injure,
        seedling_ontology.Life_Marry,
        seedling_ontology.Manufacture_Artifact,
        seedling_ontology.Movement_TransportArtifact,
        seedling_ontology.Movement_TransportPerson,
        seedling_ontology.Personnel_Elect,
        seedling_ontology.Personnel_EndPosition,
        seedling_ontology.Personnel_Nominate,
        seedling_ontology.Personnel_StartPosition,
        seedling_ontology.Transaction_Transaction,
        seedling_ontology.Transaction_TransferControl,
        seedling_ontology.Transaction_TransferMoney,
        seedling_ontology.Transaction_TransferOwnership]

    ROLES = [
        "Attacker",
        "Instrument",
        "Place",
        "Target",
        "Time",
        "Broadcaster",
        "Place",
        "Time",
        "Participant",
        "Place",
        "Participant",
        "Time",
        "Participant",
        "Affiliate",
        "Affiliation",
        "Affiliation",
        "Person",
        "Entity",
        "Sponsor",
        "Defendant",
        "Prosecutor",
        "Adjudicator",
        "Defendant",
        "Agent",
        "Instrument",
        "Victim",
        "Artifact",
        "Manufacturer",
        "Agent",
        "Artifact",
        "Destination",
        "Instrument",
        "Origin",
        "Time",
        "Agent",
        "Destination",
        "Instrument",
        "Origin",
        "Person",
        "Employee",
        "Organization",
        "Person",
        "Entity",
        "Place",
        "Beneficiary",
        "Giver",
        "Recipient",
        "Thing",
        "Time"]

if __name__ == "__main__":
    ScalingTest().run_scaling_test()
